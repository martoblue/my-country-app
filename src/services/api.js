import axios from 'axios';

const API_URL_ALL_COUNTRIES =
  'https://restcountries.com/v3.1/all?fields=name,flags';

export const getAllCountries = async () => {
  try {
    const response = await axios.get(API_URL_ALL_COUNTRIES);
    return response.data;
  } catch (error) {
    handleApiError(error, 'all countries');
    throw error;
  }
};

export const getCountryByName = async (countryName) => {
  const API_URL_COUNTRY_BY_NAME = `https://restcountries.com/v3.1/name/${countryName}?fullText=true`;
  try {
    const response = await axios.get(API_URL_COUNTRY_BY_NAME);
    console.log(response.data);
    return response.data[0];
  } catch (error) {
    handleApiError(error, countryName);
    throw error;
  }
};

const handleApiError = (error, additionalInfo) => {
  if (error.response) {
    if (error.response.status === 404) {
      console.warn(`Recibido error 404 para la solicitud ${additionalInfo}.`);
    } else {
      console.error(
        `Error en la respuesta del servidor para la solicitud ${additionalInfo}:`,
        error.response.data,
      );
    }
  } else if (error.request) {
    console.error(`No se recibi√≥ respuesta del servidor ${additionalInfo}.`);
  } else {
    console.error(
      `Error al configurar la solicitud ${additionalInfo}:`,
      error.message,
    );
  }
};
