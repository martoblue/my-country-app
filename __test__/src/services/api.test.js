import axios from 'axios';
import { getAllCountries, getCountryByName } from '@/services/api.js';
import mockCountries from './mocks/countries.mock';

jest.mock('axios');

describe('API Requests', () => {
  beforeEach(() => {
    axios.get.mockReset();
  });

  test('getAllCountries should return countries data', async () => {
    axios.get.mockResolvedValue({ data: mockCountries });

    const result = await getAllCountries();
    expect(result).toEqual(mockCountries);
    expect(axios.get).toHaveBeenCalledWith(expect.any(String));
  });

  test('getCountryByName should return country data for a valid country name', async () => {
    const mockCountryName = 'Canada';

    axios.get.mockResolvedValue({ data: [mockCountries[0]] });

    const result = await getCountryByName(mockCountryName);
    expect(result).toEqual(mockCountries[0]);
    expect(axios.get).toHaveBeenCalledWith(
      expect.stringContaining(mockCountryName),
    );
  });

  test('getCountryByName should handle 404 error gracefully', async () => {
    const mockCountryName = 'NonexistentCountry';

    axios.get.mockRejectedValue({ response: { status: 404 } });

    await expect(
      getCountryByName(mockCountryName),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
