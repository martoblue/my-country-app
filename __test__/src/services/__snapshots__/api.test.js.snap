// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`API Requests getCountryByName should handle 404 error gracefully 1`] = `undefined`;
